services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/pdf/MapGenerator.kt

```diff
@@ -0,0 +1,197 @@
+import org.geotools.renderer.lite.StreamingRenderer
+import org.locationtech.jts.geom.CoordinateFilter
+import org.locationtech.jts.geom.Polygon
+import org.locationtech.jts.io.geojson.GeoJsonReader
+import org.springframework.stereotype.Component
+
+@Component
+class MapGenerator(private val wms: WebMapServer) {
+
+    fun mapWithAreas(
+        areas: List<KaivuilmoitusAlue>,
+        hankealueet: List<SavedHankealue>,
+        imageWidth: Int,
+        imageHeight: Int,
+    ): ByteArray {
+        val bounds = calculateBounds(areas, imageWidth, imageHeight)
+
+        val layer =
+            WMSUtils.getNamedLayers(wms.capabilities).find {
+                it.title == KIINTEISTOKARTTA_LAYER_TITLE
+            }!!
+
+        val mapLayer = WMSLayer(wms, layer, KIINTEISTOKARTTA_STYLE, "image/png")
+
+        val map = MapContent()
+        map.addLayer(mapLayer)
+
+        val hankealueIds = areas.map { it.hankealueId }.toSet()
+        hankealueet
+            .filter { hankealueIds.contains(it.id) }
+            .forEach { alue ->
+                val polygons =
+                    alue.geometriat
+                        ?.featureCollection
+                        ?.features
+                        ?.map { it.geometry }
+                        ?.filterIsInstance<JsonPolygon>() ?: listOf()
+
+                val style = selectColorStyle(alue.tormaystarkasteluTulos)
+                val featureLayer = FeatureLayer(readPolygons(polygons), style)
+                map.addLayer(featureLayer)
+            }
+
+        areas
+            .flatMap { it.tyoalueet }
+            .forEach {
+                val style = selectColorStyle(it.tormaystarkasteluTulos)
+                val featureLayer = FeatureLayer(readPolygon(it.geometry), style)
+                map.addLayer(featureLayer)
+            }
+
+        val renderer: GTRenderer = StreamingRenderer()
+        renderer.mapContent = map
+
+        val envelope = bounds.asReferencedEnvelope()
+
+        // The maps are distorted, unless we request a square map. So we request a square and then
+        // crop it to fit the requested image dimensions.
+        val imageSize = max(imageWidth, imageHeight)
+        val imageBounds = Rectangle(0, 0, imageSize, imageSize)
+        val image = BufferedImage(imageBounds.width, imageBounds.height, BufferedImage.TYPE_INT_RGB)
+
+        val gr = image.createGraphics()
+        gr.paint = Color.WHITE
+        gr.fill(imageBounds)
+
+        val cropped =
+            image.getSubimage(
+                (imageSize - imageWidth) / 2,
+                (imageSize - imageHeight) / 2,
+                imageWidth,
+                imageHeight,
+            )
+
+        renderer.paint(gr, imageBounds, envelope)
+        val baos = ByteArrayOutputStream()
+        ImageIO.write(cropped, "png", baos)
+        map.dispose()
+        return baos.toByteArray()
+    }
+
+    companion object {
+        val blueish: Style by lazy { buildAreaStyle(Color(0, 98, 185)) }
```

Comment:
Is the `ish` postfix really necessary for these colours?

Reply:
Not really necessary. I wanted to distinguish them from `Color.BLUE` etc.

Reply:
I was adding an enum for the colors in the next task. I added it here already.